src/java.base/share/classes/java/lang/runtime/ObjectMethods.java

```diff
@@ -216,7 +216,17 @@ private static MethodHandle makeEquals(Class<?> receiverClass,
         MethodHandle isInstance = MethodHandles.dropArguments(CLASS_IS_INSTANCE.bindTo(receiverClass), 0, receiverClass); // (RO)Z
         MethodHandle accumulator = MethodHandles.dropArguments(TRUE, 0, receiverClass, receiverClass); // (RR)Z
 
-        for (MethodHandle getter : getters) {
+        var equalsGetters = getters.toArray(new MethodHandle[0]);
+        Arrays.sort(equalsGetters, (mh1, mh2) -> {
+            var rt1 = mh1.type().returnType();
+            var rt2 = mh2.type().returnType();
+            if (rt1.isPrimitive() == rt2.isPrimitive() || rt1.isEnum() == rt2.isEnum()) {
+                return 0;
+            }
+            return rt1.isPrimitive() || rt1.isEnum() || rt1.isArray() ? 1 : Iterable.class.isAssignableFrom(rt1) ? -1 : 0;
+        });
```

Comment:
The way this comparator is currently implemented, the `signum(c.compare(mh1, mh2)) == -signum(c.compare(mh2, mh1))` assertion won’t hold.

Reply:
This should ensure stable ordering:
```suggestion
        Arrays.sort(equalsGetters, (mh1, mh2) -> {
            var rt1 = mh1.type().returnType();
            var rt2 = mh2.type().returnType();
            return Integer.compare(
                rt1.isPrimitive() || rt1.isEnum() || rt1.isArray() ? 1 : Iterable.class.isAssignableFrom(rt1) ? -1 : 0,
                rt2.isPrimitive() || rt2.isEnum() || rt2.isArray() ? 1 : Iterable.class.isAssignableFrom(rt2) ? -1 : 0
            );
        });
```